src/parser/GenericParser.ts

```diff
@@ -33,43 +31,57 @@ export class GenericParser {
     /**
      * Parses files and calculates metrics as specified by the configuration of this {@link GenericParser} object.
      */
-    calculateMetrics() {
+    async calculateMetrics() {
         const startTime = performance.now();
 
-        const parseFiles = findFilesRecursively(
-            fs.realpathSync(this.config.sourcesPath),
-            [...grammars.keys()],
-            this.config.exclusions,
-            []
-        );
+        const parseFilesGenerator = findFilesAsync(this.config.sourcesPath, this.config.exclusions);
 
-        dlog(" --- " + parseFiles.length + " files detected\n\n");
+        const fileMetricPromises: Promise<[string, Map<string, MetricResult>]>[] = [];
+        const parseFiles: ParseFile[] = [];
 
-        let fileMetrics = new Map<string, Map<string, MetricResult>>();
         if (this.config.parseMetrics) {
             const metricsParser = new MetricCalculator(this.config);
-            fileMetrics = metricsParser.calculateMetrics(parseFiles);
+
+            for await (const file of parseFilesGenerator) {
+                fileMetricPromises.push(metricsParser.calculateMetrics(file));
```

Comment:
I'm not sure what you mean with "before Promise.all". Do you mean catching errors inside CalculateMetrics so that they are not thrown? Or surrounding Promise.all with a try-catch-block?

Reply:
```js
const promise1 = startPromise()
try {
  const promsie2 = await startPromise2()
  await promise1
} catch {}
```

This is a problem: if `promise1` is running into an error while promise2 is awaited, it is going to be detected as unhandled rejection, due to not being handled on a next tick. That is going to cause Node.js to shut down by default.

Reply:
If it is the try-catch-block, this might be resolved by the block added for handling file errors.

Reply:
No, the try catch block won't help. The `fileMetricPromises` array collects async tasks that could fail individually and the loop continues until it collected all entries. Later on we await the individual promises but that is too late. If one of them would have run into an error, it would crash as being unhandled.

Reply:
Oh, I have not refreshed that page, so my last comment does not relate to your comment before, sorry

Reply:
> ```js
> const promise1 = startPromise()
> try {
>   await startPromise2()
>   await promise1
> } catch {}
> ```
> 
> This is a problem: if `promise1` is running into an error while promise2 is awaited, it is going to be detected as unhandled rejection, due to not being handled on a next tick. That is going to cause Node.js to shut down by default.

Okay, interesting. I had searched a bit for it, but have only found that awaiting a Promise.all rejects once one of the promises rejects. I would have expected that the promise somehow "stores" the exception and that it then gets thrown when awaiting it, regardless of the order in which we await promises or something like that. Thinking of it, that may applies to the promises in C++, so that may caused my expectation.

So the only way to avoid this then is to ensure that the asynchronous function never throws an error?

Reply:
It is saved as you describe it. The issue is, that this is a frequent pattern for an incomplete promise chain. Therefore, Node.js decided to track these promises to provide users a hint that they did not handle the promise properly. Node.js just can't know what is going to happen in the future and rather fails early.

Reply:
I have now, as announced, introduced a try/catch-block in `metricsParser.calculateMetrics(...) ` to catch any error that might lead to a rejected promise in the for await of GenericParser.ts. 

For the now asynchronous loop in MetricCalculator.ts, I have used ".catch" to add a callback to handle a rejected promise, in order to avoid adding try/catch-blocks to the `calculate`-method of each metric. I hope that this also avoids the issue, as the rejection can be handled by the callback immediately?

Reply:
As this needs to be merged finally, I have opened #39 in case that my changes are still not resolving this issue.